temporal/fx.go

```diff
@@ -481,46 +481,94 @@ func ApplyClusterMetadataConfigProvider(
+					IsConnectionEnabled:      clusterInfo.Enabled,
+				}})
+		if err != nil {
+			logger.Warn("Failed to save cluster metadata.", tag.Error(err), tag.ClusterName(clusterName))
+		}
 
-	resp, err := clusterMetadataManager.GetClusterMetadata()
-	if err != nil {
-		return config.ClusterMetadata, config.Persistence, fmt.Errorf("error while fetching cluster metadata: %w", err)
-	}
+		if applied {
+			logger.Info("Successfully saved cluster metadata.", tag.ClusterName(clusterName))
+			continue
+		}
 
-	if config.ClusterMetadata.CurrentClusterName != resp.ClusterName {
-		logger.Error(
-			mismatchLogMessage,
-			tag.Key("clusterMetadata.currentClusterName"),
-			tag.IgnoredValue(config.ClusterMetadata.CurrentClusterName),
-			tag.Value(resp.ClusterName))
+		resp, err := clusterMetadataManager.GetCurrentClusterMetadata()
+		if err != nil {
+			return config.ClusterMetadata,
+				config.Persistence,
+				fmt.Errorf("error while fetching metadata from cluster %s: %w", clusterName, err)
+		}
 
-		config.ClusterMetadata.CurrentClusterName = resp.ClusterName
-	}
+		if clusterName == clusterData.CurrentClusterName {
+			if config.ClusterMetadata.CurrentClusterName != resp.ClusterName {
+				logger.Error(
+					mismatchLogMessage,
+					tag.Key("clusterMetadata.currentClusterName"),
+					tag.IgnoredValue(config.ClusterMetadata.CurrentClusterName),
+					tag.Value(resp.ClusterName))
+
+				config.ClusterMetadata.CurrentClusterName = resp.ClusterName
+			}
+			var persistedShardCount = resp.HistoryShardCount
+			if config.Persistence.NumHistoryShards != persistedShardCount {
+				logger.Error(
+					mismatchLogMessage,
+					tag.Key("persistence.numHistoryShards"),
+					tag.IgnoredValue(config.Persistence.NumHistoryShards),
+					tag.Value(persistedShardCount))
+				config.Persistence.NumHistoryShards = persistedShardCount
+			}
+		}
 
-	var persistedShardCount = resp.HistoryShardCount
-	if config.Persistence.NumHistoryShards != persistedShardCount {
-		logger.Error(
-			mismatchLogMessage,
-			tag.Key("persistence.numHistoryShards"),
-			tag.IgnoredValue(config.Persistence.NumHistoryShards),
-			tag.Value(persistedShardCount))
-		config.Persistence.NumHistoryShards = persistedShardCount
-	}
+		// Overwrite cluster information from DB
+		// TODO: load all cluster information from DB
+		if clusterInfo.RPCAddress != resp.ClusterAddress {
+			logger.Error(
+				mismatchLogMessage,
+				tag.Key("clusterInfo.rpcAddress"),
+				tag.IgnoredValue(clusterInfo.RPCAddress),
+				tag.Value(resp.ClusterAddress))
 
+			clusterInfo.RPCAddress = resp.ClusterAddress
+		}
+		if clusterInfo.InitialFailoverVersion != resp.InitialFailoverVersion {
+			logger.Error(
+				mismatchLogMessage,
+				tag.Key("clusterInfo.initialFailoverVersion"),
+				tag.IgnoredValue(clusterInfo.InitialFailoverVersion),
+				tag.Value(resp.InitialFailoverVersion))
+
+			clusterInfo.InitialFailoverVersion = resp.InitialFailoverVersion
+		}
+		if clusterInfo.Enabled != resp.IsConnectionEnabled {
```

Comment:
should check resp.IsConnectionEnabled first and skip all validation if enabled is false.

Reply:
Make sense. Will update it